x/wasm/keeper/msg_dispatcher.go

```diff
@@ -0,0 +1,174 @@
+package keeper
+
+import (
+	"github.com/CosmWasm/wasmd/x/wasm/types"
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// Messenger is an extension point for custom wasmd message handling
+type Messenger interface {
+	// DispatchMsg encodes the wasmVM message and dispatches it.
+	DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) (events []sdk.Event, data [][]byte, err error)
+}
+
+// replyer is a subset of keeper that can handle replies to submessages
+type replyer interface {
+	reply(ctx sdk.Context, contractAddress sdk.AccAddress, reply wasmvmtypes.Reply) (*sdk.Result, error)
+}
+
+// MessageDispatcher coordinates message sending and submessage reply/ state commits
+type MessageDispatcher struct {
+	messenger Messenger
+	keeper    replyer
+}
+
+// NewMessageDispatcher constructor
+func NewMessageDispatcher(messenger Messenger, keeper replyer) *MessageDispatcher {
+	return &MessageDispatcher{messenger: messenger, keeper: keeper}
+}
+
+// DispatchMessages sends all messages.
+func (d MessageDispatcher) DispatchMessages(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msgs []wasmvmtypes.CosmosMsg) error {
```

Comment:
Okay, this is just https://github.com/CosmWasm/wasmd/pull/502/files#diff-4e16010ab332946722b789b22962fb5a0814d529cd8de443b925d20ccd13705fL782-L792 without changes